script/tool/lib/src/federation_safety_check_command.dart

```diff
@@ -0,0 +1,187 @@
+          platform: platform,
+          gitDir: gitDir,
+        );
+
+  // A map of package name (as defined by the directory name of the package)
+  // to a list of changed Dart files in that package, as Posix paths relative to
+  // the package root.
+  //
+  // This only considers top-level packages, not subpackages such as example/.
+  final Map<String, List<String>> _changedDartFiles = <String, List<String>>{};
+
+  // The set of *_platform_interface packages that will have public code changes
+  // published.
+  final Set<String> _modifiedAndPublishedPlatformInterfacePackages = <String>{};
+
+  // The set of conceptual plugins (not packages) that have changes.
+  final Set<String> _changedPlugins = <String>{};
+
+  static const String _platformInterfaceSuffix = '_platform_interface';
+
+  @override
+  final String name = 'federation-safety-check';
+
+  @override
+  final String description =
+      'Checks that the change does not violate repository rules around changes '
+      'to federated plugin packages.';
+
+  @override
+  bool get hasLongOutput => false;
+
+  @override
+  Future<void> initializeRun() async {
+    final GitVersionFinder gitVersionFinder = await retrieveVersionFinder();
+    final String baseSha = await gitVersionFinder.getBaseSha();
+    print('Validating changes relative to "$baseSha"\n');
+    for (final String path in await gitVersionFinder.getChangedFiles()) {
+      // Git output always uses Posix paths.
+      final List<String> allComponents = p.posix.split(path);
+      final int packageIndex = allComponents.indexOf('packages');
+      if (packageIndex == -1) {
+        continue;
+      }
+      final List<String> relativeComponents =
+          allComponents.sublist(packageIndex + 1);
+      // The package name is either the directory directly under packages/, or
+      // the directory under that in the case of a federated plugin.
+      String packageName = relativeComponents.removeAt(0);
+      // Count the top-level plugin as changed.
+      _changedPlugins.add(packageName);
+      if (relativeComponents[0] == packageName ||
+          relativeComponents[0].startsWith('${packageName}_')) {
+        packageName = relativeComponents.removeAt(0);
+      }
+
+      if (relativeComponents.last.endsWith('.dart')) {
+        _changedDartFiles[packageName] ??= <String>[];
+        _changedDartFiles[packageName]!
+            .add(p.posix.joinAll(relativeComponents));
+      }
+
+      if (packageName.endsWith(_platformInterfaceSuffix) &&
+          relativeComponents.first == 'pubspec.yaml' &&
+          await _packageWillBePublished(path)) {
+        _modifiedAndPublishedPlatformInterfacePackages.add(packageName);
+      }
+    }
+  }
+
+  @override
+  Future<PackageResult> runForPackage(RepositoryPackage package) async {
+    if (!isFlutterPlugin(package)) {
+      return PackageResult.skip('Not a plugin.');
+    }
+
+    if (package.directory.parent.basename == 'packages') {
```

Comment:
This seems like something we might want to turn into a function since the it has very clear semantics whose details could potentially change in the future.  Something like `package.isFederated`.

Reply:
Done.